[DOC-12145] Added missing query to example 3 of WITH RECURSIVE page - Server (#196)
couchbaselabs/docs-devex: modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc

Reviewer: Hi,
Thanks for the quick fix:)
While we are at it I spotted another change required here https://github.com/couchbaselabs/docs-devex/pull/196/files#diff-edbf6e825c4644d098ef7d8b929f0f2c4097b659d5a5c49a4003b296ba30fcb5R368

OPTIONS {"levels:2"} -> OPTIONS {"levels":2}

Other that this all good.

Author: Thanks, Gaurav! Just made the small change. I'll merge this once Simon has taken a look and reviewed it :)

Reviewer 2: The Recursive SELECT syntax diagram starting at line 43 needs a `[source,ebnf]` block attribute list

Reviewer 2: The examples all get the word "Example" prepended to them automatically, so they don't need to contain the word "Example" 😄 

- "Example of simple recursive referencing" (line 143) should be just "Simple recursive referencing"
- "Example combining recursive and non-recursive CTEs with the WITH clause" (line 174) should be just "Combine recursive and non-recursive CTEs with the WITH clause"

----
====

> .Examples handling hierarchical or tree-structured data

Reviewer 2: As above, doesn't need to include the word "Example"

```suggestion
.Handle hierarchical or tree-structured data
```

====

> Before trying out these examples, you must:

Reviewer 2: This won't work until you've created the `employees` collection. I'm not 100% sure of the best place to create this. It doesn't seem to belong in the `inventory` scope, so let's put it in `tenant_agent_00`.

Best practice in the SQL++ reference is to set the query context to the bucket and scope you'll be using for the query.

```suggestion
For this example, set the query context to the `tenant_agent_00` scope in the travel sample dataset.
For more information, see xref:n1ql:n1ql-intro/queriesandresults.adoc#query-context[Query Context].
```

> . Create a dataset containing employee information:
> +

Reviewer 2: Using numbered steps with code blocks in this example creates an inconsistency, because some some code blocks have titles and some don't, and the different indents make the code blocks look wonky. Use titled code blocks throughout the example.

```suggestion
.Create a dataset containing employee information
```

[source,sqlpp]

> ----

Reviewer 2: We need to create the collection first.

```suggestion
----
CREATE COLLECTION employees IF NOT EXISTS;

```

INSERT INTO employees
   VALUES ("e1", {
       "employee_id": 1,
       "employee_name": "John"
       }),
   VALUES ("e2", {
       "employee_id": 2,
       "employee_name": "Emily",
       "manager_id": 1
       }),
   VALUES ("e3", {
       "employee_id": 3,
       "employee_name": "Mike",
       "manager_id": 1
       }),
   VALUES ("e4", {
       "employee_id": 4,
       "employee_name": "Sarah",
       "manager_id": 2
       }),
   VALUES ("e5", {
       "employee_id": 5,
       "employee_name": "Alex",
       "manager_id": 3
       }),
   VALUES ("e6", {
       "employee_id": 6,
       "employee_name": "Lisa",
       "manager_id": 2
       })
RETURNING *;
----
+

> . Create a new index:
> +

Reviewer 2: ```suggestion
.Create an index for the employees
```

[source,sqlpp]
----
CREATE INDEX e_idx ON employees(manager_id INCLUDE MISSING, employee_id, employee_name);
----

> .Example finding the level of each employee in the organization

Reviewer 2: ```suggestion
.Find the level of each employee in the organization
```

[source,sqlpp]
----
WITH recursive emplLevel AS (
   SELECT e.employee_id, e.employee_name, 0 lvl
   FROM employees e
   WHERE manager_id IS MISSING
       UNION
   SELECT e1.employee_id, e1.employee_name, e1.manager_id, emplLevel.lvl+1
   lvl
   FROM employees e1 JOIN emplLevel
   ON e1.manager_id = emplLevel.employee_id
)
SELECT * FROM emplLevel;

> ----

Reviewer 2: Show the expected results. (Feel free to cut the results down with an ellipsis `// ...` if you feel they're too long, as long as the different employee levels are clear.)

```suggestion
----

.Results
[source,json]
----
[
  {
    "emplLevel": {
      "employee_id": 1,
      "employee_name": "John",
      "lvl": 0
    }
  },
  {
    "emplLevel": {
      "employee_id": 2,
      "employee_name": "Emily",
      "lvl": 1,
      "manager_id": 1
    }
  },
  {
    "emplLevel": {
      "employee_id": 3,
      "employee_name": "Mike",
      "lvl": 1,
      "manager_id": 1
    }
  },
  {
    "emplLevel": {
      "employee_id": 4,
      "employee_name": "Sarah",
      "lvl": 2,
      "manager_id": 2
    }
  },
  {
    "emplLevel": {
      "employee_id": 6,
      "employee_name": "Lisa",
      "lvl": 2,
      "manager_id": 2
    }
  },
  {
    "emplLevel": {
      "employee_id": 5,
      "employee_name": "Alex",
      "lvl": 2,
      "manager_id": 3
    }
  }
]
----
```

> .Example resulting in the reporting hierarchy of the employees in the organization

Reviewer 2: ```suggestion
.Find the reporting hierarchy of the employees in the organization
```

[source,sqlpp]
----
SELECT e.employee_id, e.employee_name, e.manager_id,
(
    WITH RECURSIVE cte AS (
        SELECT e1.employee_id, e1.employee_name, e1.manager_id
        FROM employees e1 WHERE e1.manager_id = e.employee_id
            UNION
        SELECT e2.employee_id, e2.employee_name, e2.manager_id
        FROM employees e2, cte
        WHERE e2.manager_id = cte.employee_id
    )
    SELECT cte.* FROM cte
) as reportsTo
FROM employees e;

> ----

Reviewer 2: Show the results. (Feel free to cut the results down with an ellipsis `// ...` if you feel they're too long, as long as the different reporting structures are clear.)

```suggestion
----

.Results
[source,json]
----
[
  {
    "employee_id": 1,
    "employee_name": "John",
    "reportsTo": [
      {
        "employee_id": 2,
        "employee_name": "Emily",
        "manager_id": 1
      },
      {
        "employee_id": 3,
        "employee_name": "Mike",
        "manager_id": 1
      },
      {
        "employee_id": 4,
        "employee_name": "Sarah",
        "manager_id": 2
      },
      {
        "employee_id": 6,
        "employee_name": "Lisa",
        "manager_id": 2
      },
      {
        "employee_id": 5,
        "employee_name": "Alex",
        "manager_id": 3
      }
    ]
  },
  {
    "employee_id": 2,
    "employee_name": "Emily",
    "manager_id": 1,
    "reportsTo": [
      {
        "employee_id": 4,
        "employee_name": "Sarah",
        "manager_id": 2
      },
      {
        "employee_id": 6,
        "employee_name": "Lisa",
        "manager_id": 2
      }
    ]
  },
  {
    "employee_id": 3,
    "employee_name": "Mike",
    "manager_id": 1,
    "reportsTo": [
      {
        "employee_id": 5,
        "employee_name": "Alex",
        "manager_id": 3
      }
    ]
  },
  {
    "employee_id": 4,
    "employee_name": "Sarah",
    "manager_id": 2,
    "reportsTo": []
  },
  {
    "employee_id": 6,
    "employee_name": "Lisa",
    "manager_id": 2,
    "reportsTo": []
  },
  {
    "employee_id": 5,
    "employee_name": "Alex",
    "manager_id": 3,
    "reportsTo": []
  }
]
----
```

====

> .Examples using the OPTIONS clause and the CYCLE clause to avoid infinite recursion

Reviewer 2: ```suggestion
.Use the OPTIONS clause and the CYCLE clause to avoid infinite recursion
```

====

> Before trying out these examples, you must:

Reviewer 2: Again, let's set the query context to the bucket and scope you'll be using for the example.

```suggestion
For this example, set the query context to the `tenant_agent_00` scope in the travel sample dataset.
For more information, see xref:n1ql:n1ql-intro/queriesandresults.adoc#query-context[Query Context].
```

> . Create a dataset containing product information:
> +

Reviewer 2: Just titles for the code blocks, as above

```suggestion
.Create a dataset containing product information
```

[source,sqlpp]

> ----

Reviewer 2: Create the collection first:

```suggestion
----
CREATE COLLECTION products IF NOT EXISTS;

```

INSERT INTO products
  VALUES (uuid(), {
    "id": 1,
    "name": "Bicycle",
    "price": 299.99,
    "description": "A sturdy and reliable bicycle.",
    "category": "Sports",
    "related_items": [2, 4]
  }),
  VALUES (uuid(), {
    "id": 2,
    "name": "Helmet",
    "price": 49.99,
    "description": "A safety helmet for cycling.",
    "category": "Sports",
    "related_items": [1, 3]
  }),
  VALUES (uuid(), {
    "id": 3,
    "name": "Water Bottle",
    "price": 9.99,
    "description": "A convenient water bottle for cycling.",
    "category": "Accessories",
    "related_items": [2, 4]
  }),
  VALUES (uuid(), {
    "id": 4,
    "name": "Gloves",
    "price": 19.99,
    "description": "Comfortable gloves for cycling.",
    "category": "Sports",
    "related_items": [1, 3]
  })
RETURNING *;
----
+

> . Create a new index:
> +

Reviewer 2: ```suggestion
. Create an index for the products
```

[source,sqlpp]
----
CREATE INDEX p_idx ON products(id); 
----

> .Example resulting in infinite recursion

Reviewer 2: ```suggestion
.Query resulting in infinite recursion
```

Also, add a line _after_ this query stating what users will see if they try to run it. Something like

```asciidoc
This query will continue running without giving any results, until Couchbase Server runs out of memory.
```

  SELECT p1.*, similar_items.lvl+1 as lvl FROM products p1,
    similar_items WHERE p1.id IN similar_items.related_items
)
OPTIONS {"levels":2}
SELECT * FROM similar_items;

> ---- 

Reviewer 2: Show the expected results.

Also, add a line _after_ the query with the CYCLE clause, stating the results are the same as the previous query. You don't need to show the results twice.

```suggestion
----

.Results
[source,json]
----
[
  {
    "similar_items": {
      "category": "Sports",
      "description": "A sturdy and reliable bicycle.",
      "id": 1,
      "lvl": 0,
      "name": "Bicycle",
      "price": 299.99,
      "related_items": [
        2,
        4
      ]
    }
  },
  {
    "similar_items": {
      "category": "Sports",
      "description": "A safety helmet for cycling.",
      "id": 2,
      "lvl": 1,
      "name": "Helmet",
      "price": 49.99,
      "related_items": [
        1,
        3
      ]
    }
  },
  {
    "similar_items": {
      "category": "Sports",
      "description": "Comfortable gloves for cycling.",
      "id": 4,
      "lvl": 1,
      "name": "Gloves",
      "price": 19.99,
      "related_items": [
        1,
        3
      ]
    }
  },
  {
    "similar_items": {
      "category": "Accessories",
      "description": "A convenient water bottle for cycling.",
      "id": 3,
      "lvl": 2,
      "name": "Water Bottle",
      "price": 9.99,
      "related_items": [
        2,
        4
      ]
    }
  },
  {
    "similar_items": {
      "category": "Sports",
      "description": "A sturdy and reliable bicycle.",
      "id": 1,
      "lvl": 2,
      "name": "Bicycle",
      "price": 299.99,
      "related_items": [
        2,
        4
      ]
    }
  }
]
----
```

Reviewer 2: Another thing I've just noticed. In the reporting hierarchy example, I think the `reportsTo` alias is misleading. The query doesn't show who each employee reports to, it shows each employee's reports, i.e. the people they manage. Please can you change `reportsTo` to `reports` in this query, and in the results?

(Also, let's change some of the employee names to add some more diversity!)

SELECT * FROM similar_items;
----

> .Results

Reviewer 2: This is fine as-is, but for this one example I think it would also be okay just to say "this query will run until the Query service runs out of memory".

CYCLE id RESTRICT
SELECT * FROM similar_items;
---- 

> .Results

Reviewer 2: No need to repeat the results here: you could just have a line saying "The results are the same as for the previous query."
